api: Allow every user have multiple API keys.

YagoGG opened the pull request:
This is the first part of our effort of letting users have more than one API key associated to their account.
    
This way, they will be able to have separate keys for each client they are using, which is better from a security perspective and also allows revoking access to one of them without interfering with the others.


**Testing Plan:** lint and mypy checks passed locally, tried to fetch and regenerate API keys with the already existing endpoints.

YagoGG commented:
Provisioning master and then back to this revealed some issues that weren't there before, makes me wonder how could that happen :thinking:.

Anyways, pushed a new version with everything fixed.

YagoGG commented:
Again, I just ran `./tools/test-backend --coverage` in my dev env and everything is fine. Any ideas on why could this be happening?

The PR's branch is up to date with what I have locally.

timabbott commented:
We definitely don't want to remove the complete `test_migrations` test suite: just the existing test.  This is a test suite that is intended to be used one-off when we have a tricky migration to verify, but once that migration has been verified, we don't need the tests far in the future after that (at which point it might break).

timabbott commented:
The coverage errors are because that migrations test covered that file.  You can edit the exclude list in `test-backend` to get this passing.

In terms of cleaning up this PR, there's a bunch of places in the "big commit" where you refactor interfaces in way that could be done with the old API; I think it'll be worth splitting those out to smaller preparatory refactors.

timabbott commented:
(I reviewed and merged the first couple commits, that I'd authored).

timabbott commented on tools/test-api:
```diff
@@ -58,7 +60,7 @@ with test_server_running(force=options.force, external_host='zulipdev.com:9981')
     guest_user = do_create_user('guest@zulip.com', 'secret',
                                 get_realm('zulip'), 'Mr. Guest', 'guest')
-    api_key = guest_user.api_key
+    api_key = get_api_key(guest_user)
```
The refactor in this file (and many others) could easily be done in a preparatory commit, that introduces `get_api_key(user_profile)` as just `return user_profile.api_key`.  And then the later commit would just need to change `get_api_key`.

YagoGG commented:
Thanks for reviewing, @timabbott!

> We definitely don't want to remove the complete `test_migrations` test suite: just the existing test.

Yeah, I was expecting that commit to require some modifications. I just restored most of the stuff I removed, but now that there are no test cases in `test_migrations`, I've had to add a few files to the coverage exception list; I'm not satisfied with that, so any ideas on how can we work that out would be great :).

I'm also not sure if it's safe to remove `use_db_models`; let me know if we can and I'll make the corresponding changes.

> In terms of cleaning up this PR, there's a bunch of places in the "big commit" where you refactor interfaces in way that could be done with the old API; I think it'll be worth splitting those out to smaller preparatory refactors.

I tried to keep the commit as lean as possible, but it ended up getting quite big. I hadn't thought about the preparatory commit for `get_api_key`, thanks for the idea! I just extracted that into a separate commit, as well as a couple things more that could be done before the big one.

YagoGG commented:
I just pushed a few more commits with the new endpoints that allow creating, listing and revoking API keys.

I'm still not convinced about b36622c, so we might want to talk a bit about that before merging.

timabbott commented on zerver/views/auth.py:
```diff
@@ -156,7 +157,7 @@ def login_or_register_remote_user(request: HttpRequest, remote_username: Optiona
         # For the mobile Oauth flow, we send the API key and other
         # necessary details in a redirect to a zulip:// URI scheme.
-        api_key = get_api_key(user_profile)
+        api_key = create_user_api_key(user_profile, 'Created via REST API').api_key
```
This should definitely have a label for mobile, e.g. "Zulip mobile app".

timabbott commented on zerver/views/user_settings.py:
```diff
@@ -206,9 +206,9 @@ def delete_avatar_backend(request: HttpRequest, user_profile: UserProfile) -> Ht
     json_result = dict(
-        api_key = user_profile.api_key
+        api_key = new_api_key
     )
```
I think this refactor of the `do_regenerate_api_key` interface would be nice to split out as a preparatory commit.

timabbott commented:
Thanks for working on this @YagoGG!  It's definitely making progress.  I merged 5 of the preparatory commits, and posted a few comments.

YagoGG commented:
Thanks for reviewing, @timabbott! I just split that into a separate commit, changed the mobile API key's description as you requested, and rebased with the current master as well.

YagoGG commented:
Resolved the new merge conflicts.

thedeveloperr commented:
@YagoGG I am cleaning up this PR as asked by Tim so I might need your help regarding some doubts. Let's schedule a discussion as per your convenience. 😅

timabbott commented:
Closing in favor of #12511.
